### PR TITLE
Add Semtype test runner

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Related to #34629

### CODE REVIEW

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            relations.add(relation(sourceTypeName, targetTypeName));
+        }
+    }
+
+    private List<String> expectedSubTypeRelations(String filePath) throws IOException {
+        List<String> relations = new ArrayList<>();
+        List<String> lines = Files.readAllLines(resolvePath(filePath));
+        for (String line : lines) {
+            // This check can be changed based on other notations.
+            if (line.startsWith("// ") && line.contains("<:")) {
```
Comment (gimantha) [inline tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:116]: Shall we extract these constants to a Constant file?

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ */
+public class SemTypeTest {
+
+    @DataProvider(name = "filePathProvider")
+    public Object[] filePathProvider() {
+        File dir = resolvePath("test-src").toFile();
+        List<String> files = new ArrayList<>();
+        for (File file : Objects.requireNonNull(dir.listFiles())) {
+            // This check is used to remove failing tests.
+            if (!file.getName().contains("-f.bal")) {
```
Comment (gimantha) [inline tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:56]: Whats the purpose of this file?

Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:56]: I we found a bug in our type checker, we can put this test cases in this file, until we fix

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java (hunk 1 above)
Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java]: Changed

File: tests/jballerina-semtype-test/build.gradle [hunk 3]
```diff
@@ -0,0 +1,25 @@
+apply from: "$rootDir/gradle/javaProject.gradle"
```
Comment (pcnfernando) [inline tests/jballerina-semtype-test/build.gradle:1]: Shall we add the license header

Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/build.gradle:20]: Added

**Review by pcnfernando — APPROVED** [review]

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        Assert.assertEquals(expectedRelations.size(), 0);
+    }
+
+    private List<String> actualSubTypeRelations(String filePath) {
+        CompileResult compileResult = BCompileUtil.compile(filePath);
+        BLangPackage bLangPackage = (BLangPackage) compileResult.getAST();
+        List<BTypeDefinitionSymbol> bTypeDefinitionSymbols = new ArrayList<>();
+        for (Scope.ScopeEntry value : bLangPackage.symbol.scope.entries.values()) {
+            BSymbol bSymbol = value.symbol;
+            if (bSymbol.kind == SymbolKind.TYPE_DEF) {
```
Comment (chiranSachintha) [inline tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:84]: @KavinduZoysa We need to skip the generated type definitions
check the following example
```ballerina
// ArrayFiveOfIntFive<:ArrayOfIntFive
// ROArrayFiveOfIntFive<:ArrayOfIntFive
// ROArrayFiveOfIntFive<:ArrayFiveOfIntFive

const FIVE = 5;

type ArrayOfIntFive int[][FIVE];

type ArrayFiveOfIntFive int[FIVE][FIVE];

type ROArrayFiveOfIntFive ArrayFiveOfIntFive & readonly;

type INT int;
```

Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:84]: @chiranSachintha, what do you mean by  `generated type definitions`? 
In this example, we cannot find usages of `type INT int;`.

Comment (chiranSachintha) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:84]: for the `const FIVE = 5;` we generate type definition(with anonymous name) by the compiler. Here, don't we need to ignore those type definitions?

Comment (chiranSachintha) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:84]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/34704

**Pull request merged**